addons/mail/models/mail_message.py

```diff
@@ -948,6 +948,17 @@ def _message_format(self, fnames):
         self.check_access_rule('read')
         vals_list = self._read_format(fnames)
         safari = request and request.httprequest.user_agent.browser == 'safari'
+
+        message_ids_by_model = defaultdict(list)
+        for message in self:
+            if message.model and message.res_id:
```

Comment:
Old implementation of user notifications notably: no model, no res_id = user notification. Now there is a specific message_type.

Otherwise remember that mail.message is also used for mail.mail through inherits . You could have mail creating messages without model / res_id .